File: vs_xam/baby-flashcards/ViewLogic/FlashCardViewControllerAnimatedTransitioning.cs
using System;
using CoreGraphics;
using CoreAnimation;
using Foundation;
using UIKit;

namespace babyflashcards
{

	public class FlashCardViewControllerAnimatedTransitioning : UIViewControllerAnimatedTransitioning  {

		private IUIViewControllerContextTransitioning _transitionContext;
		private AnimDoneDelegate _animDoneDelegate;

		public bool Presenting { get; set; }

		public override double TransitionDuration (IUIViewControllerContextTransitioning transitionContext)
		{
			return 0.5;
		}

		public override void AnimateTransition (IUIViewControllerContextTransitioning transitionContext)
		{
			_transitionContext = transitionContext;

			var containerView = transitionContext.ContainerView;

			FlashCardViewController toViewController;
			UIViewController fromViewController;

			if (Presenting) {
				fromViewController = transitionContext.GetViewControllerForKey (UITransitionContext.FromViewControllerKey) as UIViewController;
				toViewController = transitionContext.GetViewControllerForKey (UITransitionContext.ToViewControllerKey) as FlashCardViewController;
			} else {
				toViewController = transitionContext.GetViewControllerForKey (UITransitionContext.FromViewControllerKey) as FlashCardViewController;
				fromViewController = transitionContext.GetViewControllerForKey (UITransitionContext.ToViewControllerKey) as UIViewController;
			}

			if(Presenting)
				containerView.AddSubview(toViewController.View);

			var originRect = toViewController.SourceFrame;

			var circleRect = new CGRect (originRect.GetMidX(), originRect.GetMidY(), 10, 10);

			var circleMaskPathInitial = UIBezierPath.FromOval(circleRect); //(ovalInRect: button.frame);
			var extremePoint = new CGPoint(circleRect.X - toViewController.View.Bounds.Width, circleRect.Y - toViewController.View.Bounds.Height ); //CGRect.GetHeight (toViewController.view.bounds));
			var radius = (float)Math.Sqrt((extremePoint.X * extremePoint.X) + (extremePoint.Y * extremePoint.Y));
			var largeCircleRect = circleRect.Inset (-radius, -radius);
			var circleMaskPathFinal = UIBezierPath.FromOval (largeCircleRect);

			CGPath fromPath;
			CGPath toPath;

			if (Presenting) {
				fromPath = circleMaskPathInitial.CGPath;
				toPath = circleMaskPathFinal.CGPath;
			} else {
				var path = new CGPath ();
				fromPath = circleMaskPathFinal.CGPath;
				toPath = circleMaskPathInitial.CGPath;
			}

			var maskLayer = new CAShapeLayer();
			maskLayer.Path = fromPath;
			if (Presenting) {
				toViewController.View.Layer.Mask = maskLayer;
			} else {
				toViewController.View.Layer.Mask = maskLayer;
			}

			var maskLayerAnimation = CABasicAnimation.FromKeyPath("path");
			maskLayerAnimation.From = ObjCRuntime.Runtime.GetNSObject(fromPath.Handle);
			maskLayerAnimation.To = ObjCRuntime.Runtime.GetNSObject(toPath.Handle);
			maskLayerAnimation.Duration = this.TransitionDuration(transitionContext);
			_animDoneDelegate = new AnimDoneDelegate (transitionContext);
			maskLayerAnimation.Delegate = _animDoneDelegate;
			maskLayer.AddAnimation(maskLayerAnimation, "path");

		}

		public override void AnimationEnded (bool transitionCompleted)
		{

		}

		private class AnimDoneDelegate : CAAnimationDelegate{

			private IUIViewControllerContextTransitioning _transitionContext;

			public AnimDoneDelegate(IUIViewControllerContextTransitioning transitionContext){
				_transitionContext = transitionContext;
			}

			public override void AnimationStopped (CAAnimation anim, bool finished)
			{
				_transitionContext.CompleteTransition (!_transitionContext.TransitionWasCancelled);
				var fromCont = _transitionContext.GetViewControllerForKey (UITransitionContext.ToViewControllerKey) as UIViewController;
				fromCont.View.Layer.Mask = null;
			}

		}

	}

}

File: xcode/Scratch1/Scratch1/FlashCardViewControllerAnimatedTransitioning.swift
//
//  FlashCardViewControllerAnimatedTransitioning.swift
//  Scratch1
//
//  Created by Michael Ishmael on 21/05/2016.
//  Copyright © 2016 66Bytes. All rights reserved.
//

import Foundation
import CoreGraphics;
import UIKit;

public class FlashCardViewTransitioningDelegate : NSObject, UIViewControllerTransitioningDelegate
{
    
    public func animationControllerForPresentedController(presented: UIViewController, presentingController presenting: UIViewController, sourceController source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return FlashCardViewControllerAnimatedTransitioning(presenting: true)
    }
    
}


public class FlashCardViewControllerAnimatedTransitioning : NSObject, UIViewControllerAnimatedTransitioning  {
    
    private var _transitionContext:UIViewControllerContextTransitioning?;
    private var  _animDoneDelegate:AnimDoneDelegate?;
    
    public var presenting:Bool
    
    public init(presenting:Bool) {
        self.presenting = presenting;
    }
    
    
    @objc public func transitionDuration(transitionContext: UIViewControllerContextTransitioning?) -> NSTimeInterval {
        return NSTimeInterval(0.5)
    }
    
    
    @objc public func animateTransition(transitionContext: UIViewControllerContextTransitioning) {
        
        _transitionContext = transitionContext;
    
        let containerView = transitionContext.containerView();
    
        let toViewController:FlashCardViewController?;
        //let fromViewController:UIViewController;
    
        if (self.presenting) {
				//fromViewController = transitionContext.viewControllerForKey(UITransitionContextFromViewKey)!   as UIViewController;
				toViewController = transitionContext.viewControllerForKey (UITransitionContextToViewKey) as? FlashCardViewController;
        } else {
                    toViewController = transitionContext.viewControllerForKey(UITransitionContextFromViewKey) as? FlashCardViewController;
                 //   fromViewController = transitionContext.viewControllerForKey(UITransitionContextToViewControllerKey)! as UIViewController;
        }
        
        if toViewController == nil {
            return
        }
    
        if self.presenting{
            containerView!.addSubview(toViewController!.view);
        }
				
    
    let originRect = toViewController!.sourceFrame;
    let circleRect = CGRect(x: originRect.midX, y: originRect.midY, width: 10, height: 10);
    
    let circleMaskPathInitial = UIBezierPath.init(ovalInRect: circleRect); //(ovalInRect: button.frame);
    let extremePoint = CGPoint(x: circleRect.minX - toViewController!.view.bounds.width, y:circleRect.minY - toViewController!.view.bounds.height ); //CGRect.GetHeight (toViewController.view.bounds));
    let radius = Float(sqrt((extremePoint.x * extremePoint.x) + (extremePoint.y * extremePoint.y)));
    let largeCircleRect = circleRect.insetBy(dx: CGFloat(-radius), dy: CGFloat(-radius))
    let circleMaskPathFinal = UIBezierPath.init(ovalInRect: largeCircleRect)
        
    var fromPath:CGPath
    var toPath:CGPath
    
    if (presenting) {
        fromPath = circleMaskPathInitial.CGPath;
        toPath = circleMaskPathFinal.CGPath;
    } else {
        fromPath = circleMaskPathFinal.CGPath;
        toPath = circleMaskPathInitial.CGPath;
    }
    
    let maskLayer = CAShapeLayer();
    maskLayer.path = fromPath;
    if (presenting) {
        toViewController!.view.layer.mask = maskLayer;
    } else {
        toViewController!.view.layer.mask = maskLayer;
    }
    
    let maskLayerAnimation = CABasicAnimation.init(keyPath: "path")
    maskLayerAnimation.fromValue = fromPath;
    maskLayerAnimation.toValue = toPath;
    maskLayerAnimation.duration = self.transitionDuration(transitionContext);
    _animDoneDelegate = AnimDoneDelegate.init(transitionContext: transitionContext);
    maskLayerAnimation.delegate = _animDoneDelegate;
    maskLayer.addAnimation(maskLayerAnimation, forKey: "path");
    
    }
    
    @objc public func animationEnded(transitionCompleted: Bool) {
    
    }
    
}

class AnimDoneDelegate : CAAnimation {
    
    private let _transitionContext:UIViewControllerContextTransitioning
    
    init(transitionContext:UIViewControllerContextTransitioning){
        _transitionContext = transitionContext;
        super.init()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
     func animationStopped ( anim:CAAnimation, finished:Bool)
    {
        _transitionContext.completeTransition (!_transitionContext.transitionWasCancelled());
        let fromCont = _transitionContext.viewControllerForKey (UITransitionContextToViewControllerKey)! as UIViewController;
    fromCont.view.layer.mask = nil;
    }
    
}



